### PR TITLE
PYTHON-3698 Support mypy 1.2 --strict testing (part 1)

### PR DESCRIPTION
**Pull request by blink1073**
I broke this up into about half of the errors to save review.

### CODE REVIEW

File: mypy.ini [hunk 1]
```diff
@@ -1,4 +1,5 @@
 [mypy]
+python_version = 3.7
```
Comment (blink1073) [inline mypy.ini:2]: This setting enables us to test against Python 3.7 typing semantics directly, simplifying local running and CI.

Comment (ShaneHarvey) [reply mypy.ini:2]: Does this mean that we now _never_ test against 3.11?

File: pymongo/encryption.py [hunk 2]
```diff
@@ -923,7 +941,7 @@ def get_key(self, id: Binary) -> Optional[RawBSONDocument]:
         """
         self._check_closed()
         assert self._key_vault_coll is not None
-        return self._key_vault_coll.find_one({"_id": id})
+        return cast(RawBSONDocument, self._key_vault_coll.find_one({"_id": id}))
```
Comment (ShaneHarvey) [inline pymongo/encryption.py:944]: Can we type _key_vault_coll as `Collection[RawBSONDocument]` rather than adding these casts?

File: pymongo/change_stream.py [hunk 3]
```diff
@@ -422,17 +422,17 @@ def try_next(self) -> Optional[_DocumentType]:
         self._start_at_operation_time = None
 
         if self._decode_custom:
-            return _bson_to_dict(change.raw, self._orig_codec_options)
-        return change
+            return cast(_DocumentType, _bson_to_dict(change.raw, self._orig_codec_options))
+        return cast(_DocumentType, change)
```
Comment (ShaneHarvey) [inline pymongo/change_stream.py:426]: Can we avoid these casts by defining _orig_codec_options in terms of _DocumentType? casts are bad in my opinion because they can mask legitimate typing bugs so we should only use them as a last resort.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: mypy.ini (hunk 1 above)
Comment (blink1073) [reply mypy.ini:2]: It's in the example config from `mypy`, I think of it like the version we use in `black`.    Unless there is a backwards-incompatible change to typing, this should be fine.

File: pymongo/encryption.py (hunk 2 above)
Comment (blink1073) [reply pymongo/encryption.py:944]: Done

File: pymongo/change_stream.py (hunk 3 above)
Comment (blink1073) [reply pymongo/change_stream.py]: Fixed.  However, I noticed that `with_options` should defer to the `CodecOptions` type, since that is often the option you are changing.  This requires casting the parent type.

File: mypy.ini (hunk 1 above)
Comment (ShaneHarvey) [reply mypy.ini:2]: Okay sounds good to me.

File: pymongo/collection.py [hunk 4]
```diff
@@ -431,7 +431,7 @@ def with_options(
         read_preference: Optional[_ServerMode] = None,
         write_concern: Optional[WriteConcern] = None,
         read_concern: Optional[ReadConcern] = None,
-    ) -> Collection[_DocumentType]:
+    ) -> Collection[_DocumentTypeArg]:
```
Comment (ShaneHarvey) [inline pymongo/collection.py:434]: Could we revert this change and discuss it in a new jira ticket? I think the interaction of Collection and CodecOption's DocumentType is not well defined so I don't think we should make breaking changes like this without more consideration of the whole api.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: pymongo/collection.py (hunk 4 above)
Comment (blink1073) [reply pymongo/collection.py:434]: Yeah, I agree there isn't a good way to handle this with the typing system.  I think we should stick with what we've previously shipped.

**Review by ShaneHarvey — APPROVED** [review]
